### PR TITLE
Shrink the ZFS closure by making mail support optional

### LINKED ISSUE #83865: zfs pulls in mailutils, mariadb-connector and openssl-dev
**Describe the bug**

On a ZFS-enabled 20.03 system:

```
$ nix why-depends /run/current-system /nix/store/6ibqafpi22dkhdd1pb8d35z1kxvri5gp-openssl-1.1.1d-dev
/nix/store/ffchggmyb568733jqhd95y63r9knbhj3-nixos-system-nixos-20.03.20200328.b88ff46
└───activate: …bf2b22m-setup-etc.pl /nix/store/s5h0c7dzapaqd37mr226p3mag4z8smy4-etc/etc...if (( _localstatus > …
    → /nix/store/s5h0c7dzapaqd37mr226p3mag4z8smy4-etc
    └───etc/zfs/zed.d/zed.rc -> /nix/store/mzqv3glb7nrsjkdxz1hm9rdna8ns5nd4-etc-zed.rc
        → /nix/store/mzqv3glb7nrsjkdxz1hm9rdna8ns5nd4-etc-zed.rc
        └───/: …ZED_EMAIL_PROG="/nix/store/dirpf6jyhqhlk6vh0kv0g1h2c9mqbbqb-mailutils-3.8/bin/mail".…
            → /nix/store/dirpf6jyhqhlk6vh0kv0g1h2c9mqbbqb-mailutils-3.8
            └───bin/comsatd: …qb-mailutils-3.8/lib:/nix/store/16d91vilik4pj8qg10nd7zp1yzyp715p-mariadb-connector-c-3.1.5/lib/m…
                → /nix/store/16d91vilik4pj8qg10nd7zp1yzyp715p-mariadb-connector-c-3.1.5
                └───nix-support/propagated-build-inputs: …7pwa-curl-7.68.0-dev /nix/store/6ibqafpi22dkhdd1pb8d35z1kxvri5gp-openssl-1.1.1d-dev /ni>
                    → /nix/store/6ibqafpi22dkhdd1pb8d35z1kxvri5gp-openssl-1.1.1d-dev
lines 1-10/10 (END)
```

IMHO zed does not need the ability to send email by default. Also mariadb-connector should have a separate -dev output so that its `propagated-build-inputs` doesn't end up in the system closure.

### PR DESCRIPTION
**Pull request by cole-h**
###### Motivation for this change

Fixes https://github.com/NixOS/nixpkgs/issues/83865.

```sh
# in a nixpkgs checkout
$ nix why-depends -f. zfs mariadb-connector-c
'/nix/store/7s4kkijwv4wd3sl6jgn20yjai8zvini6-zfs-user-2.0.0' does not depend on '/nix/store/1h8r94mfhlfqd2wxqqnzff5rhjbrpimf-mariadb-connector-c-3.1.8'

$ nix why-depends -f. zfs openssl.dev
'/nix/store/7s4kkijwv4wd3sl6jgn20yjai8zvini6-zfs-user-2.0.0' does not depend on '/nix/store/vrsxvbphmqnhz8fww6zv0fa5fi8xipy3-openssl-1.1.1i-dev'

$ nix why-depends -f. zfs mailutils
'/nix/store/7s4kkijwv4wd3sl6jgn20yjai8zvini6-zfs-user-2.0.0' does not depend on '/nix/store/yjj734x773ydfkz1qx1z640nni1gd9bm-mailutils-3.10'

# in my system flake
$ nixos-rebuild build --flake .#bootstrap &>/dev/null && nix why-depends ./result nixpkgs#openssl.dev
'/nix/store/y4mjgvnvrpjhkiz4kkla2m4pckmb399x-nixos-system-scadrial-21.03.20210113.dirty-cosmere' does not depend on '/nix/store/vrsxvbphmqnhz8fww6zv0fa5fi8xipy3-openssl-1.1.1i-dev
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

There's probably a better way to do this, but at least this gets the ball rolling / starts the conversation. Pinged all the people who participated in the original issue.

**TODO:** write docs?

### CODE REVIEW

**Comment by cole-h** [comment]
@ofborg test zfs

File: pkgs/tools/system/smartmontools/default.nix [hunk 1]
```diff
@@ -26,7 +27,7 @@ in stdenv.mkDerivation rec {
   postPatch = "cp -v ${driverdb} drivedb.h";
 
   configureFlags = [
-    "--with-scriptpath=${stdenv.lib.makeBinPath [ mailutils inetutils ]}"
+    "--with-scriptpath=${lib.makeBinPath ([ inetutils ] ++ lib.optional enableMail mailutils)}"
```
Comment (hmenke) [inline pkgs/tools/system/smartmontools/default.nix:30]: Does it even need `inetutils` when not enabling mail support?  I can't think of another reason why smartmontools should talk to the network.

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 2]
```diff
@@ -431,7 +435,7 @@ in
       };
 
       services.zfs.zed.settings = {
-        ZED_EMAIL_PROG = mkDefault "${pkgs.mailutils}/bin/mail";
+        ZED_EMAIL_PROG = mkIf cfgZED.enableMail (mkDefault "${pkgs.mailutils}/bin/mail");
```
Comment (hmenke) [inline nixos/modules/tasks/filesystems/zfs.nix:438]: This also has to somehow override the `zfs` package with `enableMail`.

**Review by hmenke — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/zfs/default.nix [hunk 3]
```diff
@@ -8,8 +8,8 @@
 , libtirpc
 , nfs-utils
 , gawk, gnugrep, gnused, systemd
-, smartmontools, sysstat, sudo
-, pkgconfig
+, smartmontools, enableMail ? false
+, sysstat, sudo, pkgconfig
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/zfs/default.nix:12]:
```suggestion
, sysstat, sudo, pkg-config
```

File: pkgs/tools/system/smartmontools/default.nix [hunk 4]
```diff
@@ -1,5 +1,6 @@
 { lib, stdenv, fetchurl, autoreconfHook
-, mailutils, inetutils
+, mailutils, enableMail ? true
+, inetutils
 , IOKit ? null , ApplicationServices ? null }:
```
Comment (SuperSandro2000) [inline pkgs/tools/system/smartmontools/default.nix:4]:
```suggestion
, IOKit, ApplicationServices }:
```

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 2 above)
Comment (jonringer) [reply nixos/modules/tasks/filesystems/zfs.nix:438]: maybe we can add enableMail to `passthru` and just reference that.

File: pkgs/tools/system/smartmontools/default.nix (hunk 1 above)
Comment (cole-h) [reply pkgs/tools/system/smartmontools/default.nix:30]: The only `inetutils` binary it uses in `update-smart-drivedb` is `ftp`, which has a comment nearby  of `# OpenBSD`. Say the word and I'll make them both optional; I don't know if that single use of `ftp` is worth "breaking" or not (considering it also supports `curl`, `wget`, `svn`, etc).

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 2 above)
Comment (cole-h) [reply nixos/modules/tasks/filesystems/zfs.nix:444]: Done in 01f6aee23cf6f464e2b300bf274fc31ce408d65f and b87934f6cf97d2a2eec2aa7c7fe26eb87061629d.

Comment (cole-h) [reply nixos/modules/tasks/filesystems/zfs.nix:444]: Wait, I forgot to override zfs (lol), another commit incoming.

Comment (cole-h) [reply nixos/modules/tasks/filesystems/zfs.nix:444]: Should I go the assertion approach, or the override approach? I feel they're mutually exclusive.

**Comment by cole-h** [comment]
This is ready for review (basically just squashed everything). I'll write docs after the contents are approved to limit the amount of (potential) rewriting.

File: pkgs/tools/system/smartmontools/default.nix [hunk 5]
```diff
@@ -1,5 +1,6 @@
 { lib, stdenv, fetchurl, autoreconfHook
-, mailutils, inetutils
+, mailutils, enableMail ? true
```
Comment (SuperSandro2000) [inline pkgs/tools/system/smartmontools/default.nix:2]:
```suggestion
, enableMail ? true, mailutils
```

File: pkgs/os-specific/linux/zfs/default.nix [hunk 6]
```diff
@@ -8,8 +8,8 @@
 , libtirpc
 , nfs-utils
 , gawk, gnugrep, gnused, systemd
-, smartmontools, sysstat, sudo
-, pkg-config
+, smartmontools, enableMail ? false
+, sysstat, sudo, pkg-config
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/zfs/default.nix:12]: It is not clear for me which packages belong to enableMail.

Edit:

Maybe like this if we only need smartmontools.

```suggestion
, sysstat, sudo, pkg-config
, enableMail ? false, smartmontools
```

**Review by doronbehar — APPROVED** [review]
LGTM though I haven't tested a thing.

**Pull request merged**